api/models/ssl.go

```diff
@@ -273,14 +234,136 @@ func ListSSLs(a string) (SSLs, error) {
 	return ssls, nil
 }
 
+func UpdateSSL(app, process string, port int, body, key string, chain string) (*SSL, error) {
+	a, err := GetApp(app)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// validate app is not currently updating
+	if a.Status != "running" {
+		return nil, fmt.Errorf("can not update app with status: %s", a.Status)
+	}
+
+	// store old cert name
+	oldCertName := certName(app, process, port)
+
+	// validate process exists
+	if oldCertName == "" {
+		return nil, fmt.Errorf("no certificate configured for %s port %d", process, port)
+	}
+
+	outputs := a.Outputs
+
+	balancer := outputs[fmt.Sprintf("%sPort%dBalancerName", UpperName(process), port)]
+
+	if balancer == "" {
+		return nil, fmt.Errorf("Balancer ouptut not found. Please redeploy your app and try again.")
+	}
+
+	// upload new cert
+	arn, err := uploadCert(a, process, port, body, key, chain)
+
+	if err != nil {
+		return nil, err
+	}
+
+	input := &elb.SetLoadBalancerListenerSSLCertificateInput{
+		LoadBalancerName: aws.String(balancer),
+		LoadBalancerPort: aws.Int64(int64(port)),
+		SSLCertificateId: aws.String(arn),
+	}
+
+	// change cert on listener
+	operation := func() error {
+		_, err = ELB().SetLoadBalancerListenerSSLCertificate(input)
+
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	err = backoff.Retry(operation, backoff.NewExponentialBackOff())
+
+	if err != nil {
+		return nil, err
+	}
+
+	// delete old cert
+	operation = func() error { return deleteCert(oldCertName) }
+
+	go backoff.Retry(operation, backoff.NewExponentialBackOff())
+
+	// update cloudformation
+	req := &cloudformation.UpdateStackInput{
+		StackName:           aws.String(a.Name),
+		Capabilities:        []*string{aws.String("CAPABILITY_IAM")},
+		UsePreviousTemplate: aws.Bool(true),
+	}
+
+	params := a.Parameters
+
+	params[fmt.Sprintf("%sPort%dCertificate", UpperName(process), port)] = arn
+
+	for key, val := range params {
+		req.Parameters = append(req.Parameters, &cloudformation.Parameter{
+			ParameterKey:   aws.String(key),
+			ParameterValue: aws.String(val),
+		})
+	}
+
+	_, err = CloudFormation().UpdateStack(req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ssl := SSL{
+		Port:    port,
+		Process: process,
+	}
+
+	return &ssl, nil
```

Comment:
In my testing the whole update process took 3 or 4 seconds. I think it could be a little longer on a balancer that was serving heavy traffic and had to flush all the requests associated with the old cert before allowing a delete.